core/src/main/java/com/linecorp/armeria/common/stream/FuseableStreamMessage.java

```diff
@@ -0,0 +1,307 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static com.linecorp.armeria.common.stream.StreamMessageUtil.EMPTY_OPTIONS;
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import javax.annotation.Nullable;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.common.stream.FuseableStreamMessage.MapperFunction.Type;
+
+import io.netty.util.concurrent.EventExecutor;
+
+final class FuseableStreamMessage<T, U> implements StreamMessage<U> {
+
+    static <T> FuseableStreamMessage<T, T> of(StreamMessage<? extends T> source,
+                                              Predicate<? super T> predicate) {
+        return new FuseableStreamMessage<>(source, MapperFunction.of(predicate));
+    }
+
+    static <T, R> FuseableStreamMessage<T, R> of(StreamMessage<? extends T> source,
+                                                 Function<? super T, ? extends R> function) {
+        return new FuseableStreamMessage<>(source, MapperFunction.of(function));
+    }
+
+    private final StreamMessage<Object> source;
```

Comment:
```suggestion
    private final StreamMessage<T> source;
```

Reply:
Unfortunately, we cannot statically know the element type of `source`. 
For example:
```
// In this case, `T` and `U` types are Integer.
StreamMessage<Integer> filtered = StreamMessage.of(1, ...).filter(x -> x > 10);

// In this case, `T` type is `Integer` and `U` type is `String`
StreamMessage<String> mapped = filtered.map(x -> x.toString);

// In this case, `T` type is `String` and `U` type is `Boolean`
// However, the type of `mapped.source` should be `StreamMessage<Integer>` which publishes the original item.
StreamMessage<Boolean> mapped = mapped.map(x -> x.equals(...));
```

Reply:
https://github.com/line/armeria/pull/3351/commits/43f73214db67fa4ff9541e5908bdf64105e7f8d5
I think that was the reason that made me confused. 😄

Reply:
Sorry, it was my mistake. 😅

Reply:
Maybe add some comments on why it's not `T`?